### DISCUSSION TITLE
Reuse prefetched next/link cache on hover

### DISCUSSION THREAD

**Discussion by fitiskin** [vercel/next.js#40268] [Ideas]
### Describe the feature you'd like to request

Idea: reduce the HTTP requests count that `next/link` generates when hovering over a link.

![next-link-prefetch-on-hover](https://user-images.githubusercontent.com/5507542/188568972-a3d1f15b-ba10-4b8c-8888-c2c3911398a1.gif)

My thoughts:
- skip prefetching on hover when the link has already been prefetched (prefetch=true/undefined on viewport)
- do not repeat prefetch on hover after the previous one
- do not call prefetch on hover when the same prefetch request is pending
- there should be a [way to disable it](https://github.com/vercel/next.js/pull/37299) completely

### Describe the solution you'd like

Respect next/link [prefetched](https://github.com/vercel/next.js/blob/77e7529dfe493081a84b95c27ebaa1bc265d2396/packages/next/client/link.tsx#L68) cache on hover.

### Describe alternatives you've considered

* throttling
* [hover duration](https://github.com/vercel/next.js/pull/37299#issuecomment-1206745702)
* other cache than next/link `prefetched`

**Comment by jettlove** [comment]
agree, nobody wants to DDOS their backend

**Comment by SukkaW** [comment]
I am wondering if this is the intended behavior. Imagine that the user hovers over the link, but not perform the navigate. And then the prefetched data get outdated. So once the user hover again, the data should be re-prefetched by then.

**Comment by fitiskin** [comment]
Maybe. I wasn't sure if this was not intentional behavior.

I considered this as a feature because of the last point, which changes the behavior of the component and in my opinion needs confirmation from maintainers.

**Comment by SukkaW** [comment]
> Maybe. I wasn't sure if this was not intentional behavior. I can try to make a fix by myself since I already started yesterday.
> 
> I considered this as a feature because of the last point, which changes the behavior of the component and in my opinion needs confirmation from maintainers.

Actually, I am considering introducing a 10ms/20ms delay before prefetching (for onMouseEnter only, onTouchStart doesn't need that). This should prevent unwanted prefetches.

**Comment by fitiskin** [comment]
> Actually, I am considering introducing a 10ms/20ms delay before prefetching (for onMouseEnter only, onTouchStart doesn't need that).

@trompx has a [similar idea](https://github.com/vercel/next.js/pull/37299#issuecomment-1206745702)

**Comment by ijjk** [comment]
Hi, as mentioned by @SukkaW this is intentional since with ISR the content could have been updated since it was prefetched due to being in the viewport so we attempt an additional optimistic prefetch on hover to allow getting fresh content. In most cases this will probably be a 304 which is a very minimal request (the payload body is not sent) and not expensive for the server to handle. 

If you aren't leveraging ISR and know the content won't change and would prefer stale content always be used you can disable this under an experimental config already `experimental.optimisticClientCache: false` https://github.com/vercel/next.js/blob/fa08a17768cd6934d7d1ef944849c54a162ee799/packages/next/server/config-shared.ts#L80

**Reply by meotimdihia** [reply]
This feature should be documented.

**Reply by khuezy** [reply]
@ijjk What about the case where the server gets throttled if you instantaneously spam it 10 requests b/c you have 10 Links in the viewport?
It would be better to stagger the requests instead of spamming all at once.

**Comment by Pontus91** [comment]
Any news regarding this?

**Comment by SukkaW** [comment]
I am about to wake my previous PR https://github.com/vercel/next.js/pull/40310 up (the PR is currently conflicting with the canary branch, I will start working on it ASAP).

**Comment by kvnol** [comment]
I disable the prefetch on all links because my server is tired. 😂

**Comment by vamcs** [comment]
If you're on Next 13 and you have a middleware, this may be the middleware prefetching data (especially if the returned data is an empty object). Check this thread out: https://github.com/vercel/next.js/discussions/44596

**Comment by yannickbuntsma** [comment]
I see that in the network tab these requests have status code 304 (Not Modified). Quoting [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304): 
> The HTTP 304 Not Modified client redirection response code indicates that there is _no need to retransmit the requested resources_. It is an implicit _redirection to a cached resource_. 

If I'm reading that right and my knowledge is on par, this means that the browser is serving this data from its cache and the server will never receive this request. Based on the HTTP caching rules in-browser, the browser will know when to consider any data in its cache to be stale and refetch the link accordingly.

**Reply by daveycodez** [reply]
My Vercel logs are saying otherwise

**Comment by ogheo** [comment]
In nextjs 15 you can't navigate to an already visited page due to optimisticClientCache,
so solved like this for now, maybe it could help someone else:
```
  experimental: {
    optimisticClientCache: false,
    staleTimes: {
      dynamic: 0,
      static: 0,
    },
  },
```
